### PR TITLE
8309808: BytecodeTracer prints wrong BSM for invokedynamic

### PR DESCRIPTION
**Pull request by iklam**
This PR fixes the tracing of invokedynamic bytecodes:

```
before: 0x00007f3aa8400385  5 invokedynamic 1 <java/lang/Object.<init>()V>
after:  0x00007f9bcc4056a5  5 invokedynamic bsm=37 13 <makeConcatWithConstants(I)Ljava/lang/String;>
```

I've also added WhiteBox functionality to make it easy to write new tests for the BytecodeTracer.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8309808](https://bugs.openjdk.org/browse/JDK-8309808): BytecodeTracer prints wrong BSM for invokedynamic (**Bug** - P4)

### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [5a173e59](https://git.openjdk.org/jdk/pull/14429/files/5a173e5938921aaba0941535b52a39e5baf5c89d)
 * [Matias Saavedra Silva](https://openjdk.org/census#matsaave) (@matias9927 - Committer) ⚠️ Review applies to [5a173e59](https://git.openjdk.org/jdk/pull/14429/files/5a173e5938921aaba0941535b52a39e5baf5c89d)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14429/head:pull/14429` \
`$ git checkout pull/14429`

Update a local copy of the PR: \
`$ git checkout pull/14429` \
`$ git pull https://git.openjdk.org/jdk.git pull/14429/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14429`

View PR using the GUI difftool: \
`$ git pr show -t 14429`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14429.diff">https://git.openjdk.org/jdk/pull/14429.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14429#issuecomment-1588168098)

### CODE REVIEW

**Review by coleenp — APPROVED** [review]
Ok, looks good to me.

File: src/hotspot/share/interpreter/bytecodeTracer.cpp [hunk 1]
```diff
@@ -302,10 +302,16 @@ bool BytecodePrinter::check_obj_index(int i, int& cp_index, outputStream* st) {
...
-  assert(ConstantPool::is_invokedynamic_index(i), "not secondary index?");
-  i = ConstantPool::decode_invokedynamic_index(i) + ConstantPool::CPCACHE_INDEX_TAG;
-
-  return check_cp_cache_index(i, cp_index, st);
+  ConstantPool* constants = _current_method->constants();
+  if (constants->cache() == nullptr) {
+    cp_index = i; // TODO: This is wrong on little-endian. See JDK-8309811.
+  } else {
+    assert(ConstantPool::is_invokedynamic_index(i), "not secondary index?");
+    i = ConstantPool::decode_invokedynamic_index(i);
```
Comment (matias9927) [inline src/hotspot/share/interpreter/bytecodeTracer.cpp:310]: This could be changed to int indy_index = ... instead of reassigning i

**Review by matias9927 — APPROVED** [review]
Looks good, thanks for this! Just one nit:

File: src/hotspot/share/interpreter/bytecodeTracer.cpp (hunk 1 above)
Comment (iklam) [reply src/hotspot/share/interpreter/bytecodeTracer.cpp]: Thanks. I've added an `indy_index` variable.

**Comment by iklam** [comment]
Thanks @coleenp and @matias9927 for the review.
/integrate

**Pull request closed without merging**